cluster: loss-userspace cluster-deploy fails primary reassert on master — blocks every owed dataplane smoke, and its failures look like HA regressions

## The shared loss-userspace cluster cannot complete `cluster-deploy`, on master

Found while running the smoke owed by #7230. **Reproduced on clean `origin/master` with no local changes**, so it is the environment, not a branch.

```
ERROR: post-deploy primary reassert FAILED: node0 is not primary for every
redundancy group after 15 verification attempts. The cluster is left in
whatever state the transfer produced; do NOT run an HA smoke against it and
read the failure as an HA regression.

Redundancy group: 1     node0  200  secondary
  Takeover ready: no (userspace XSK liveness not proven)
Redundancy group: 2     node0  200  primary
  Takeover ready: no (userspace XSK liveness not proven)

make: *** [Makefile:555: cluster-deploy] Error 1
```

`node0` is **secondary for RG1** after a deploy that is supposed to reassert it as primary for every group, and **XSK liveness is unproven on both RG1 and RG2**.

## Why this needs its own issue

**Every lane that owes a dataplane smoke is currently blocked**, and blocked in a way that produces *plausible-looking* failures rather than an obvious outage. Running `test-connectivity` against the cluster in this state yields:

```
Results: 14 passed, 5 failed, 1 skipped
  - cluster: fw0 → fw1 heartbeat (10.99.0.2)
  - cluster: fw1 → fw0 heartbeat (10.99.0.1)
  - cluster: fw0 → fw1 fabric (10.99.1.2)
  - cluster: fw1 → fw0 fabric (10.99.1.1)
  - cluster: LAN host → internet IPv6
```

Those read exactly like an HA regression in the branch under test. **The deploy tooling already anticipates this** — its own error says *"do NOT run an HA smoke against it and read the failure as an HA regression"* — but that warning is emitted by the **deploy**, and a lane that runs the smoke as a separate step, or reads only the smoke's summary, never sees it.

So the failure mode is: a lane attributes five cluster-link failures to its own change, and either chases them or weakens something to make them pass.

## Determined by measurement, not inference

The branch that hit this (`#7230`, WireGuard keepalive attribution) touches only `userspace-dp/src/afxdp/wg/*` and `coordinator/wg_control/dispatch.rs` — no HA, VRRP, session-sync or fabric code. That is an argument. The measurement is that a **detached worktree at clean `origin/master` (`daba95969`) fails the same deploy the same way.**

## Not attempting repair

The cluster is shared (#1875) and repair here is destructive. Reporting rather than reaching for it, and I did not `rm` or force the lock — both runs went through `with-cluster.sh` and released cleanly.

Related: #7673 (failover throughput on the same cluster), which may share a cause.

--- a/_Log.md
+++ b/_Log.md
@@ -107071,3 +107071,13 @@ prose edit above them added. No diff falls in the new test body.
   - **File(s)**: pkg/daemon/external_hostname_watch_6863.go, pkg/daemon/mgmt_listener_reassert.go, pkg/daemon/management.go, pkg/daemon/daemon.go
 
 <!-- LOG-CLOSED-SENTINEL-6874: nothing may be appended below this line. Write to docs/log/<issue>.md instead. -->
+
+- **Timestamp**: 2026-08-27 02:35
+  - **Action**: #7688 — clear the PEER's manual-failover pin in the post-deploy
+    primary reassert. ManualFailover is local, unsynced state, so a pin left on
+    node1 by an earlier smoke was invisible to node0 and blocked the election
+    the reassert depends on, failing `cluster-deploy` as a plausible-looking HA
+    regression. Verified by measurement: node1 read `Manual=yes` on all 3 RGs;
+    resetting it there made `cluster-deploy` on clean master pass rc=0.
+  - **File(s)**: test/incus/deploy-lib.sh, test/incus/cluster-setup.sh,
+    test/incus/deploy-lib-selftest.sh

--- a/test/incus/cluster-setup.sh
+++ b/test/incus/cluster-setup.sh
@@ -727,7 +727,7 @@ rolling_detect_secondary() {
 # unrelated smoke target's preflight, where the first hypothesis is that the
 # change under test broke HA.
 reassert_primary_node0() {
-	deploy_reassert_primary_node0 "$(r "$VM0")"
+	deploy_reassert_primary_node0 "$(r "$VM0")" "$(r "$VM1")"
 }
 
 # deploy_rolling_deb sequences the deb cut across both nodes (secondary

--- a/test/incus/deploy-lib-selftest.sh
+++ b/test/incus/deploy-lib-selftest.sh
@@ -263,6 +263,12 @@ _fake_cli_status() {
 
 reset_cli_mock() {
 	CLI_LOG=$(mktemp)
+	# $CLI_LOG records the command TEXT ONLY: incus() shifts the instance away
+	# before _fake_exec sees it, so a reset issued to the peer is byte-identical
+	# there to a duplicate reset issued to node0. $FAKE_CALL_LOG keeps the full
+	# argv, so it is the only log that can answer WHICH NODE a command went to
+	# -- which is the whole property under test in #7688.
+	FAKE_CALL_LOG=$(mktemp)
 	CLI_IDX_FILE=$(mktemp)
 	printf '0' >"$CLI_IDX_FILE"
 	CLI_REQUEST_RC=0
@@ -829,6 +835,43 @@ test_reassert_issues_reset_transfer_reset_per_rg() {
 	fi
 }
 
+test_reassert_clears_the_peer_manual_pin() {
+	reset_cli_mock
+	CLI_STATUS_RESPONSES=("$STATUS_BOTH_RG_NODE0_PRIMARY")
+	( deploy_reassert_primary_node0 "fake:vm0" "fake:vm1" ) >/dev/null 2>&1 || true
+	# Assert on the INSTANCE-bearing log, not $CLI_LOG. Keyed on $CLI_LOG this
+	# cell would stay green with the peer reset deleted, because node0 is reset
+	# twice per RG anyway and the texts are identical.
+	local peer node0
+	peer=$(grep -c '^exec fake:vm1 -- cli -c request chassis cluster failover reset redundancy-group' "$FAKE_CALL_LOG" || true)
+	node0=$(grep -c '^exec fake:vm0 -- cli -c request chassis cluster failover reset redundancy-group' "$FAKE_CALL_LOG" || true)
+	if (( peer == 2 && node0 == 4 )); then
+		ok "reassert: clears the PEER's manual pin once per RG (peer=$peer, node0=$node0)"
+	else
+		bad "reassert: peer resets=$peer (want 2, one per RG), node0 resets=$node0 (want 4).
+ManualFailover is local, unsynced state: a pin left on node1 is invisible to
+node0's CLI and blocks the election the transfer depends on (#7688)."
+	fi
+}
+
+test_reassert_call_site_passes_the_peer() {
+	# Bind the WIRING, not just the function. The peer is an OPTIONAL second
+	# argument -- deliberately, so the other reassert cells can keep driving the
+	# single-node form -- which means a call site that omits it silently gets
+	# the pre-#7688 one-sided behaviour while every other test here still
+	# passes. Only reading the production call site can catch that.
+	local f="${SCRIPT_DIR}/cluster-setup.sh"
+	local call
+	call=$(grep '^[[:space:]]*deploy_reassert_primary_node0 ' "$f" | head -1)
+	if [[ "$call" == *'VM0'*'VM1'* ]]; then
+		ok "reassert: cluster-setup.sh call site passes BOTH nodes"
+	else
+		bad "reassert: cluster-setup.sh passes no peer instance, so the peer pin
+is never cleared and the deploy still fails its primary reassert (#7688).
+  got: $call"
+	fi
+}
+
 test_reassert_transfer_rc_is_not_the_verdict() {
 	reset_cli_mock
 	# Every request returns non-zero, but the cluster ends up correct. The
@@ -1016,6 +1059,8 @@ test_reassert_dies_when_role_not_achieved
 test_reassert_issues_reset_transfer_reset_per_rg
 test_reassert_transfer_rc_is_not_the_verdict
 test_reassert_is_wired_into_both_deploy_paths
+test_reassert_clears_the_peer_manual_pin
+test_reassert_call_site_passes_the_peer
 test_ownership_verdict_ok
 test_ownership_verdict_diverged
 test_ownership_verdict_nostream

--- a/test/incus/deploy-lib.sh
+++ b/test/incus/deploy-lib.sh
@@ -324,6 +324,10 @@ DEPLOY_REASSERT_VERIFY_DELAY="${DEPLOY_REASSERT_VERIFY_DELAY:-2}"
 # instead of handing the next smoke an inverted cluster.
 deploy_reassert_primary_node0() {
 	local rinst="$1"
+	# #7688: the PEER instance. ManualFailover is LOCAL, UNSYNCED node state --
+	# there is no `manual` field in the proto and none in the heartbeat sync --
+	# so a pin left on the peer cannot be observed or cleared from node0.
+	local pinst="${2:-}"
 	local status rgs rg try
 
 	status=""
@@ -340,6 +344,19 @@ deploy_reassert_primary_node0() {
 
 	while read -r rg; do
 		[[ -n "$rg" ]] || continue
+		# Clear the PEER's manual pin FIRST. A pinned peer cannot participate
+		# in the election this transfer depends on -- test-failover.sh says it
+		# outright: "ManualFailover blocks election even when the peer is
+		# lost." Because the flag is local and unsynced, resetting it on node0
+		# (below) does nothing for a pin left on node1 by an earlier smoke:
+		# every HA smoke resets BOTH nodes as a preamble but none as a
+		# teardown, so the cluster is routinely left pinned on exit. The next
+		# deploy is then what fails, and it fails as "node0 is not primary for
+		# every redundancy group" -- indistinguishable from an HA regression in
+		# whatever branch happened to deploy next (#7688).
+		if [[ -n "$pinst" ]]; then
+			incus exec "$pinst" -- cli -c "request chassis cluster failover reset redundancy-group $rg" >/dev/null 2>&1 || true
+		fi
 		incus exec "$rinst" -- cli -c "request chassis cluster failover reset redundancy-group $rg" >/dev/null 2>&1 || true
 		incus exec "$rinst" -- cli -c "request chassis cluster failover redundancy-group $rg node 0" >/dev/null 2>&1 || true
 		incus exec "$rinst" -- cli -c "request chassis cluster failover reset redundancy-group $rg" >/dev/null 2>&1 || true

